developer/src/common/web/utils/src/xml-utils.ts

```diff
@@ -123,11 +125,54 @@ const GENERATOR_OPTIONS: KeymanXMLOptionsBag = {
   },
 };
 
+export interface LineColumn {
+  line: number;
+  column?: number;
+}
+
 /** wrapper for XML parsing support */
 export class KeymanXMLReader {
   public constructor(public type: KeymanXMLType) {
   }
 
+  public static offsetToLineColumn(offset: number, text: string) : LineColumn {
+    const lines: string[] = text.split("\n");
+    for (let line = 1; line < lines.length + 1; line++) { // 1-based
+      const linestr = lines[line - 1];
+      if (linestr.length < offset) {
+        offset = offset - (linestr.length + 1); // count newline at end
+        continue;
+      }
+      return { line, column: offset };
+    }
+    return { line: lines.length + 1 };
+  }
```

Comment:
This is quite inefficient to do for every message; can we do the parse just once when loading the file and store line lengths in a more efficient manner with an R-tree or kd-tree or the like?

Reply:
right. actually, all it probably needs is an array of numbers:  `const lens : number[] = text.split("\n").map(l => l.length+1)`

Reply:
> actually, all it probably needs is an array of numbers

The problem with an array is that the whole array has to be scanned to find the matching line. Faster than scanning the file again, but still a lot slower than a R-tree or like which is optimized for nearest-match

Reply:
The current PR only scans one time per file. Could we look at optimizing the lookup separately? I'll encapsulate the lookup separately so that it can be replaced without disruption.